File: algorithms/c++/bubble-sort.cpp
/* 
* Bubble sort
* Time complexity - O(n^2)
*/

#include <iostream>

void sort(int *arr, int size)
{
    for (int i = 0; i < size; i++)
    {
        bool swapped = false;
        for (int j = 1; j < size - i; j++)
        {
            if (arr[j - 1] > arr[j])
            {
                std::swap(arr[j - 1], arr[j]);
                swapped = true;
            }
        }
        if (!swapped)
            break;
    }
}

void print(int *arr, int size)
{
    std::cout << "{ ";
    for (int i = 0; i < size; i++)
    {
        std::cout << arr[i] << ", ";
    }
    std::cout << "\b\b }" << std::endl;
}

int main()
{
    int arr[] = {56, 78, 11, 23, 8, 12};
    int size = sizeof(arr) / sizeof(arr[0]);
    sort(arr, size);
    print(arr, size);
    return 0;
}

File: algorithms/js/bubble-sort.js
/* 
* Bubble sort
* Time complexity - O(n^2)
*/

const sort = function (arr, compare) {
    for (let i = 0; i < arr.length; i++) {
        let swapped = false;
        for (let j = 1; j < arr.length - i; j++) {
            if (compare(arr[j - 1], arr[j])) {
                [arr[j], arr[j - 1]] = [arr[j - 1], arr[j]];
                swapped = true;
            }
        }
        if (!swapped) break;
    }
    return arr;
}

// driver code

const arr = [56, 78, 11, 23, 8, 12];
console.log(sort(arr, (a, b) => a < b));
console.log(sort(arr, (a, b) => a > b));
